File: ios/Sources/MetaAdsPlugin/MetaAdsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(MetaAdsPlugin)
public class MetaAdsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MetaAdsPlugin"
    public let jsName = "MetaAds"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadRewardedVideo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showRewardedVideo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRewardedVideoLoaded", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadInterstitial", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showInterstitial", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isInterstitialLoaded", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setTestMode", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addTestDevice", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = MetaAds()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let appId = call.getString("appId") else {
            call.reject("App ID is required")
            return
        }
        
        let testMode = call.getBool("testMode") ?? false
        
        implementation.initialize(appId: appId, testMode: testMode) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error ?? "Unknown initialization error")
            }
        }
    }
    
    @objc func loadRewardedVideo(_ call: CAPPluginCall) {
        guard let placementId = call.getString("placementId") else {
            call.reject("Placement ID is required")
            return
        }
        
        implementation.loadRewardedVideo(placementId: placementId) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error ?? "Unknown error loading rewarded video")
            }
        }
    }
    
    @objc func showRewardedVideo(_ call: CAPPluginCall) {
        implementation.showRewardedVideo { success, reward in
            var result: [String: Any] = ["success": success]
            if let reward = reward {
                result["reward"] = reward
            }
            call.resolve(result)
        }
    }
    
    @objc func isRewardedVideoLoaded(_ call: CAPPluginCall) {
        call.resolve(["loaded": false])
    }
    
    @objc func loadInterstitial(_ call: CAPPluginCall) {
        guard let placementId = call.getString("placementId") else {
            call.reject("Placement ID is required")
            return
        }
        
        implementation.loadInterstitial(placementId: placementId) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error ?? "Unknown error loading interstitial")
            }
        }
    }
    
    @objc func showInterstitial(_ call: CAPPluginCall) {
        implementation.showInterstitial { success in
            call.resolve(["success": success])
        }
    }
    
    @objc func isInterstitialLoaded(_ call: CAPPluginCall) {
        call.resolve(["loaded": false])
    }
    
    @objc func setTestMode(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? false
        implementation.setTestMode(enabled: enabled)
        call.resolve()
    }
    
    @objc func addTestDevice(_ call: CAPPluginCall) {
        if let deviceId = call.getString("deviceId") {
            implementation.addTestDevice(deviceId: deviceId)
        }
        call.resolve()
    }
}

File: android/src/main/java/com/eliazavatta/plugins/metaads/MetaAdsPlugin.java
package com.eliazavatta.plugins.metaads;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "MetaAds")
public class MetaAdsPlugin extends Plugin {

    private MetaAds implementation = new MetaAds();

    @PluginMethod
    public void initialize(PluginCall call) {
        String appId = call.getString("appId");
        Boolean testMode = call.getBoolean("testMode", false);

        if (appId == null) {
            call.reject("App ID is required");
            return;
        }

        implementation.initialize(
            getContext(),
            appId,
            testMode,
            new MetaAds.InitializationCallback() {
                @Override
                public void onSuccess() {
                    call.resolve();
                }

                @Override
                public void onError(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void loadRewardedVideo(PluginCall call) {
        String placementId = call.getString("placementId");

        if (placementId == null) {
            call.reject("Placement ID is required");
            return;
        }

        implementation.loadRewardedVideo(
            getContext(),
            placementId,
            new MetaAds.AdLoadCallback() {
                @Override
                public void onAdLoaded() {
                    call.resolve();
                }

                @Override
                public void onError(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void showRewardedVideo(PluginCall call) {
        implementation.showRewardedVideo(
            new MetaAds.RewardedVideoCallback() {
                @Override
                public void onAdShown() {
                    // Ad shown successfully
                }

                @Override
                public void onRewardEarned(String rewardType, int rewardAmount) {
                    JSObject ret = new JSObject();
                    ret.put("success", true);
                    JSObject reward = new JSObject();
                    reward.put("type", rewardType);
                    reward.put("amount", rewardAmount);
                    ret.put("reward", reward);
                    call.resolve(ret);
                }

                @Override
                public void onAdClosed() {
                    // Ad closed without reward
                    JSObject ret = new JSObject();
                    ret.put("success", false);
                    call.resolve(ret);
                }

                @Override
                public void onError(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void isRewardedVideoLoaded(PluginCall call) {
        boolean loaded = implementation.isRewardedVideoLoaded();
        JSObject ret = new JSObject();
        ret.put("loaded", loaded);
        call.resolve(ret);
    }

    @PluginMethod
    public void loadInterstitial(PluginCall call) {
        String placementId = call.getString("placementId");

        if (placementId == null) {
            call.reject("Placement ID is required");
            return;
        }

        implementation.loadInterstitial(
            getContext(),
            placementId,
            new MetaAds.AdLoadCallback() {
                @Override
                public void onAdLoaded() {
                    call.resolve();
                }

                @Override
                public void onError(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void showInterstitial(PluginCall call) {
        implementation.showInterstitial(
            new MetaAds.InterstitialCallback() {
                @Override
                public void onAdShown() {
                    JSObject ret = new JSObject();
                    ret.put("success", true);
                    call.resolve(ret);
                }

                @Override
                public void onAdClosed() {
                    // Ad closed
                }

                @Override
                public void onError(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void isInterstitialLoaded(PluginCall call) {
        boolean loaded = implementation.isInterstitialLoaded();
        JSObject ret = new JSObject();
        ret.put("loaded", loaded);
        call.resolve(ret);
    }

    @PluginMethod
    public void setTestMode(PluginCall call) {
        Boolean enabled = call.getBoolean("enabled", false);
        implementation.setTestMode(enabled);
        call.resolve();
    }

    @PluginMethod
    public void addTestDevice(PluginCall call) {
        String deviceId = call.getString("deviceId");
        if (deviceId != null) {
            implementation.addTestDevice(deviceId);
        }
        call.resolve();
    }
}
